### ISSUE TITLE
`beartype_this_package`, PEP 526 checker & class methods

### ISSUE THREAD

**Issue by MaximilienLC** [beartype/beartype#293]
Hi Cecil,

Love the package, love it even more since `beartype.claw.beartype_this_package()`.

Small question: should the PEP 526 checker work inside class methods? For both regular and instance variables? I've been trying to get this to work but just want to confirm.

For instance, with the following three files:

`main.py`
```
from package import TestClass, test_func

if __name__ == "__main__":
    TestClass()
    test_func()
```

`package/__init__.py`
```
from beartype import BeartypeConf
from beartype.claw import beartype_this_package

beartype_this_package(
    conf=BeartypeConf(
        is_pep484_tower=True,
        is_color=True,
        is_debug=True,
    ),
)

from package.test import TestClass, test_func

__all__ = ["TestClass", "test_func"]
```

`test.py`
```
class TestClass:
    def __init__(self) -> None:
        self.a: int = "hi"


def test_func():
    b: int = "hi"
```

Calling `test_func()` will raise an exception but `TestClass()` won't.

Am I doing this wrong?

Best,
A fellow Ontarian.

**Comment by leycec** [comment]
Go, go... **Ontario!**

![it's happening](https://media.tenor.com/M4wNTynB5X8AAAAd/cheers-cheerleader.gif)

> Small question: should the PEP 526 checker work inside class methods?

_\*gulp\*_

Small answer: "Erhm... yes. I failed to test something again, didn't I?"

On the one hand, I failed to test something again. On the other hand, I know exactly why this is happening.

This is happening because `beartype.claw` selectively disables PEP 562-based type-checking within the bodies of classes. What does this mean? Allow me to now tiresomely bore us all. Python standardized class field types *before* Python standardized PEP 562, which means that older existing PEP standards actually violate PEP 562. One great example, if you're still curious and have yet to pass out on your keyboard:

```python
@dataclass
class MuhDataclass(object):
    # This annotated field declaration is *NOT* safely
    # type-checkable by "beartype.claw". Why? Because a
    # dataclass "field" instance is *NOT* a valid integer and
    # thus violates the type hint annotating this field. Since
    # PEP 681 standardizes declarations like this as
    # semantically valid, @beartype has *NO* alternative but
    # to quietly turn a blind eye to what otherwise might be
    # considered a type violation.
    muh_unsafe_field: int = field(default=0xCAFEBABE)
```

So, Python's most popular PEPs (like [PEP 557 – Data Classes](https://peps.python.org/pep-0557) and [PEP 681 – Data Class Transforms](https://peps.python.org/pep-0681)) violate Python's *other* most popular PEPs (like PEP 562). We can't exactly ignore or disable dataclasses. But we *can* ignore PEP 562. Not often and we really shouldn't, of course. But what you gonna do?

So, @beartype "resolves" this and similar PEP conflicts by just disabling PEP 562-based type-checking within the bodies of classes. It is horribly saddening. It is Python standards.

## I Passed Out on My Keyboard

**Yup.** Called it. I would've done the same.

The **tl;dr** here is that I *will* fix this catastrophic issue "shortly" and push out a new stable @beartype 0.16.4 release for certain optimistic definitions of "shortly." The mid-week fatigue we all feel is real.

**Comment by leycec** [comment]
**Resolved by 0a96480.** @beartype 0.16.4 lands this Friday evening, including this fix and improved (albeit still imperfect) Python 3.12 compatibility. *Iiiiiiiiiiincoming!*

![sad mcsadface](https://media.tenor.com/_7TBOA1GxO4AAAAd/fail-hurricane.gif)

**Comment by MaximilienLC** [comment]
Hehe thanks a lot for the (very) quick update,

Just tried it out and it indeed now type checks variables inside static class methods, but will regular and instance variables inside regular class methods ever be type-checkable? Not sure if it's tied in with your dataclass example.

By the way, I think you mixed up PEP 562 & 526 hhh, blame it on me, raising issues at 9pm 🙄

**Comment by leycec** [comment]
**Gah!** Issues are multiplying before my very eyes. Unsurprisingly, you're *absolutely* right. Python's abstract syntax tree (AST) API appears to semantically represent:

* Non-attribute annotated variable assignments like `o: int = "ho"` different from...
* Attribute annotated variable assignments like `self.a: int = "hi"`.

Thanks so much for pointing out all of our painful failures. Let's reopen this for the good of LLaMa-kind. My face is exhaling. :face_exhaling:

**Comment by leycec** [comment]
**Hah, hah.** Possibly *finally* resolved by 84613f2. I'd like you to decide whether @beartype finally nailed these edge cases to the floor, @MaximilienLC. Would you mind testing this one final time on your end?

This is surprisingly non-trivial functionality. In fact, I'm convinced *no* other Python libraries currently do this. I googled. Oh, how I googled. And... I found nothing. @beartype is all there is. Go, go, typing bear! :bear:

**Comment by MaximilienLC** [comment]
Passes all tests on my side! You're the best @leycec!
